zebra-chain/src/parallel/tree.rs

```diff
@@ -160,36 +191,116 @@ impl NoteCommitmentTrees {
     }
 
     /// Update the sapling note commitment tree.
+    #[allow(clippy::unwrap_in_result)]
     fn update_sapling_note_commitment_tree(
         mut sapling: Arc<sapling::tree::NoteCommitmentTree>,
         sapling_note_commitments: Vec<sapling::tree::NoteCommitmentUpdate>,
-    ) -> Result<Arc<sapling::tree::NoteCommitmentTree>, NoteCommitmentTreeError> {
+    ) -> Result<
+        (
+            Arc<sapling::tree::NoteCommitmentTree>,
+            Option<(sapling::tree::Node, bool, u16)>,
+        ),
+        NoteCommitmentTreeError,
+    > {
         let sapling_nct = Arc::make_mut(&mut sapling);
+        let distance_to_complete_subtree =
+            (1 << TRACKED_SUBTREE_HEIGHT) - sapling_nct.count() % (1 << TRACKED_SUBTREE_HEIGHT);
+        let should_find_subtree_roots =
+            sapling_note_commitments.len() as u64 >= distance_to_complete_subtree;
+
+        // MAX_BLOCK_BYTES is less than 2^21 bytes, notes are 2^5 bytes, and a full subtree has
+        // 2^16 notes. Blocks should not contain multiple level 16 roots.
+        let mut subtree_root = None;
+
+        if should_find_subtree_roots {
+            let starts_new_subtree =
+                sapling_note_commitments.len() as u64 != distance_to_complete_subtree;
+            let complete_subtree_index = distance_to_complete_subtree as usize - 1;
```

Comment:
How about `complete_subtree_note_index`?

Reply:
Sounds good. Or we could even compute the value directly in the `if` condition since we don't use the value anywhere else?

Reply:
> Or we could even compute the value directly in the if condition since we don't use it anywhere else?

Assuming the compiler will optimize the subtraction, and do it only once.

Reply:
Renamed in https://github.com/ZcashFoundation/zebra/pull/7334/commits/70387762886e8af443972ee5629363fbdac52b70.